### PR TITLE
[FLINK-18848][python] Fix to_pandas to handle retraction data properly

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request fix to_pandas to handle retraction data properly*

## Brief change log

  - *fix to_pandas to handle retraction data properly*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added tests test_to_pandas_for_retract_table*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/table/runtime/arrow/ArrowUtils.java [hunk 1]
```diff
@@ -611,6 +617,21 @@ private static void readFully(ReadableByteChannel channel, ByteBuffer dst) throw
...
+		if (isStreamingMode(table)) {
+			StreamTableEnvironment tableEnv = (StreamTableEnvironment) ((TableImpl) table).getTableEnvironment();
+			try {
+				tableEnv.toAppendStream(table, Row.class);
+			} catch (Throwable t) {
+				if (t.getMessage().contains("toAppendStream doesn't support consuming update changes") ||
+						t.getMessage().contains("Table is not an append-only table")) {
+					isRetractTable = true;
+				} else {
+					throw new RuntimeException("Failed to determine whether the given table is append only", t);
```
Comment (WeiZhong94) [inline flink-python/src/main/java/org/apache/flink/table/runtime/arrow/ArrowUtils.java:630]: Add a period at the end?

File: flink-python/src/main/java/org/apache/flink/table/runtime/arrow/ArrowUtils.java [hunk 2]
```diff
@@ -679,6 +707,24 @@ public boolean hasNext() {
...
+		while (data.hasNext()) {
+			Row element = data.next();
+			if (element.getKind() == RowKind.INSERT || element.getKind() == RowKind.UPDATE_AFTER) {
+				element.setKind(RowKind.INSERT);
+				result.add(element);
+			} else {
+				element.setKind(RowKind.INSERT);
+				if (!result.remove(element)) {
+					throw new RuntimeException(
+						String.format("Could not remove element '%s', should never happen", element));
```
Comment (WeiZhong94) [inline flink-python/src/main/java/org/apache/flink/table/runtime/arrow/ArrowUtils.java:721]: ditto

File: flink-python/src/main/java/org/apache/flink/table/runtime/arrow/ArrowUtils.java [hunk 3]
```diff
@@ -692,6 +738,22 @@ private static boolean isBlinkPlanner(Table table) {
...
+			return false;
+		} else if (tableEnv instanceof StreamTableEnvironment) {
+			return true;
+		} else if (tableEnv instanceof TableEnvironmentImpl) {
+			java.lang.reflect.Field isStreamingModeMethod = TableEnvironmentImpl.class.getDeclaredField("isStreamingMode");
+			isStreamingModeMethod.setAccessible(true);
+			return (boolean) isStreamingModeMethod.get(tableEnv);
+		} else {
+			throw new RuntimeException(String.format(
+				"Could not determine the streaming mode for table environment class %s", tableEnv.getClass()));
```
Comment (WeiZhong94) [inline flink-python/src/main/java/org/apache/flink/table/runtime/arrow/ArrowUtils.java:753]: ditto

**Comment by dianfu** [comment]
@WeiZhong94 Thanks for the review. I will address the comments when merging the PR.

**Pull request closed without merging**